BackEnd/source/server.js

```diff
@@ -0,0 +1,257 @@
+}
+
+function getLeftDiagnoalPoint(columnIndex, rowIndex, gameBoardArray) { // PURE FUNCTION
+  let columnPos = columnIndex;
+  let rowPos = rowIndex;
+  let lowestLeftDiagonal;
+
+  let foundEnd = false;
+  while (foundEnd === false) {
+    if (columnPos - 1 < 0 || rowPos + 1 > gameBoardArray[0].length - 1) {
+      lowestLeftDiagonal = [columnPos, rowPos];
+      foundEnd = true;
+    }
+    columnPos -= 1;
+    rowPos += 1;
+  }
+  return lowestLeftDiagonal;
+}
+
+function checkLeftDiagonalWin(colAndRowIndex, gameBoardArray) { // PURE FUNCTION
+  let fourInARowCount = 0;
+
+  let columnPos = colAndRowIndex[0];
+  let rowPos = colAndRowIndex[1];
+  while (columnPos >= 3 && rowPos >= 3) {
+    fourInARowCount = 0;
+    for (let i = 1; i < 4; i += 1) {
+      if (gameBoardArray[columnPos - i][rowPos - i] !== null) {
+        if (gameBoardArray[columnPos][rowPos] === gameBoardArray[columnPos - i][rowPos - i]) {
+          fourInARowCount += 1;
+        } else {
+          break;
+        }
+      } else {
+        break;
+      }
+    }
+    if (fourInARowCount >= 3) {
+      return gameBoardArray[columnPos][rowPos];
+    }
+    columnPos -= 1;
+    rowPos -= 1;
+  }
+  return null;
+}
+
+function getRightDiagnoalPoint(columnIndex, rowIndex, gameBoardArray) { // PURE FUNCTION
+  let columnPos = columnIndex;
+  let rowPos = rowIndex;
+  let lowestRightDiagonal;
+
+  let foundEnd = false;
+  while (foundEnd === false) {
+    if (columnPos + 1 > gameBoardArray.length - 1 || rowPos + 1 > gameBoardArray[0].length - 1) {
+      lowestRightDiagonal = [columnPos, rowPos];
+      foundEnd = true;
+    }
+    columnPos += 1;
+    rowPos += 1;
+  }
+  return lowestRightDiagonal;
+}
+
+function checkWinner(columnIndex, rowIndex) {
+  const lowestLeftDiagPoint = getLeftDiagnoalPoint(columnIndex, rowIndex, state.gameBoardValues);
+  const lowestRightDiagPoint = getRightDiagnoalPoint(columnIndex, rowIndex, state.gameBoardValues);
+
+  if (checkColumnWin(columnIndex, state.gameBoardValues) != null) {
+    return checkColumnWin(columnIndex, state.gameBoardValues);
+  } if (checkRowWin(rowIndex, state.gameBoardValues) != null) {
+    return checkRowWin(rowIndex, state.gameBoardValues);
+  } if (checkRightDiagonalWin(lowestLeftDiagPoint, state.gameBoardValues) != null) {
+    return checkRightDiagonalWin(lowestLeftDiagPoint, state.gameBoardValues);
+  } if (checkLeftDiagonalWin(lowestRightDiagPoint, state.gameBoardValues) != null) {
+    return checkLeftDiagonalWin(lowestRightDiagPoint, state.gameBoardValues);
+  }
+  return null;
+}
+
+app.use(express.static('./FrontEnd/source'));
+app.use(express.json());
+
+app.post('/setup', (req, res) => {
+  updateStateValues(req.body);
+  state.gameBoardValues = setupGameBoardValuesArray(state.gameBoardRows, state.gameBoardColumns);
+  res.json({
+    result: state,
+  });
+});
+
+app.post('/placeCounter', (req, res) => {
+  res.json({
+    updatedState: state,
+    placedCounterPos: placeCounter(req.body.column),
```

Comment:
This works, but what it's doing is updating the state that is being put into the object on the line above. To me, line 226 conceptually suggests '`updatedState` is the current state object', rather than '`updatedState` is an object that will be made valid by the time I finish building this object. 

So although it's longer and works just the same, I'd rather see
```js
const placedPos = placeCounter(req.body.column);
res.json({
    updatedState: state,
    placedCounterPos: placedPos,
})
```
as this is much easier to understand in my opinion